Convert lexer from Linked Linked list to Array based allocation

benjones opened the pull request:
Benefits:

* Fewer allocations (1 per pool vs 1 per token)
* Better locality
* Smaller token struct size(no next pointer)
* Lots of functions that take pointers can take refs now instead
* Eliminate some tricky linked list code and replace it with clear range style code
* I think a similar design could be used for Loc objects (nodes/tokens would store a LocManager* and an index, 2 indices to have a LocRange)

Drawbacks

* Huge diff... not sure if that can be avoided, fortunately the parser is well covered by tests
* Implementation leaks info about the lexer.  Needed because ASM statements are parsed during semantic.  This could be improved if the parser was changed to operate on TokenRanges rather than call nextToken, etc.  This issue is also why Parser instances were changed from scope to auto.  Again, making the rest of the parser TokenRange aware would avoid the need for this.
* Extra variables needed in a few places because ref functions can't accept temporaries
* The choice of Pool parameters were pulled out of thin air... would probably be good to have something more principled
* a couple of API names seem bit cumbersome (makeRangeFromHere)

benjones committed:
convert lexer to use array based rather than linked list based allocation scheme.  Reduces allocations and simplifies some tricky code

rikkimax commented:
Can you please split out the memory ownership logic from the lexer class?
I.e. its own struct, that is never copied and is assumed to only have the one instance inside the lexer class.

Its out of scope of this PR, but it would be good to split out the tokenization from the lexer class as well. So that the class only acts as a accumulator of all the behavior and ties them together.

thewilsonator commented:
> I think a similar design could be used for Loc objects

OMG, yes please! Doing that well would cut a massive amount of memory from the compiler.

> Huge diff... not sure if that can be avoided

Not so much of a problem. I'm sure reviewers will tell you how to break it up if need be.

> Implementation leaks info about the lexer. 

This is unfortunate. Since only ASM needs it, would it be possible to copy a block of tokens and make ASM use an array rather than a range of token.

benjones commented:
I thought of giving the range a "detach" method that copies out the tokens or something, but the ASM parsing calls parseExpression, so I don't think it would work.

benjones commented:
@rikkimax like a TokenPool struct?  I think each lexer would need its own, but yeah, that's an easy change.

rikkimax commented:
@benjones Yeah I think you got the gist of what I was thinking about.

It would be nice to make that class a little simpler.
It may not be one of Walter's goals any time soon, but it would be nice to get that module ready to be split up.

benjones committed:
pull token pool allocator into its own struct

benjones commented:
Looks like the Windows_OMF x64 test is running out of memory.  Any suggestions on how to fix that?

Otherwise, I think I addressed the suggestions that don't need to go in a subsequent PR, and it's ready for a more thorough review

WalterBright commented on src/dmd/lexer.d:
```diff
@@ -202,6 +202,129 @@ unittest
+
+    ///Token pool looks like a big array
+    public ref Token getByIndex(size_t index){
+        if (index >= PoolSize*poolsAllocated)
```
Operators are preceded and succeeded by a single space. This is erratic throughout these changes.

WalterBright commented on src/dmd/lexer.d:
```diff
@@ -202,6 +202,129 @@ unittest
+        assert(poolsNeeded == poolsAllocated + 1); //don't alloc more than 1 pool at a time
+        if (poolsNeeded > pools.length)
+        {
+            const newLength = pools.length > 0 ? 2*pools.length : DefaultPoolListSize;
```
The new size should be a factor of 1.5, not 2. Compute it by multiplying by 3 then dividing by 2. Round up.

WalterBright commented on src/dmd/lexer.d:
```diff
@@ -202,6 +202,129 @@ unittest
+private:
+    enum PoolSize = 1024; //number of tokens per pool
+    enum DefaultPoolListSize = 64; //number of pools to allocate by default
+    Token[][] pools;
```
Just use an `Array` - the code here is re-inventing what `Array` does.

benjones replied:
I chose this representation to avoid copying tokens, but I'm going to experiment to see if a 1D array is actually better.  For now, I updated this to use Array!(Token[])

WalterBright commented on src/dmd/lexer.d:
```diff
@@ -202,6 +202,129 @@ unittest
+ * Pool allocator for Token objects, used by the Lexer.
+ * Tokens are never freed so Token pointers will never change
+ */
+private struct TokenPool{
```
Please follow normal DMD bracing format.

WalterBright commented on src/dmd/lexer.d:
```diff
@@ -202,6 +202,129 @@ unittest
+{
+private:
+    Lexer lex; //where are these tokens coming from?
+    size_t start_, stop_;
```
Please don't use _ in identifier names unless it is a postfix for a keyword name.

WalterBright commented on src/dmd/lexer.d:
```diff
@@ -202,6 +202,129 @@ unittest
+
+public:
+    nothrow:
+    /// returns the first token in the range
```
Since these are standard functions, there's no need to document them unless they do something non-standard.

WalterBright commented on src/dmd/lexer.d:
```diff
@@ -202,6 +202,129 @@ unittest
+        return lex.getByIndex(start_ +1);
+    }
+    /// add the next token in the scanner to the back of the range
+    Token growBack()
```
Group all the standard range functions together, don't intermingle with other functions.

WalterBright commented on src/dmd/iasmgcc.d:
```diff
@@ -284,27 +285,22 @@ Ldone:
-    Token *toklist = null;
-    Token **ptoklist = &toklist;
+    TokenRange toklist = s.tokens;
+    auto p = cast(Parser!ASTCodegen)toklist.getLexer;
```
Don't call a lexer a `p`.

WalterBright replied:
Call it `lex` or `lexer`.

benjones replied:
It's being used as a parser because it has to call parseExpression later on.    I don't love this design but couldn't come up with a simpler alternative

WalterBright commented on src/dmd/lexer.d:
```diff
@@ -202,6 +202,129 @@ unittest
+    {
+        return stop_ - start_;
+    }
+    /// return the index of the first token in the lexer's array of tokens
```
Use Ddoc style comments, i.e. here use `Returns:`

WalterBright commented on src/dmd/lexer.d:
```diff
@@ -300,62 +427,60 @@ class Lexer
+    final void seekTo(size_t index)
+    {
+        nextTokenIndex = index;
+        token = getByIndex(nextTokenIndex++);
```
This will be more efficient written as:
```
token = getByIndex(index);
nextTokenIndex = index + 1;
```

WalterBright commented:
The DMD parser is highly sensitive to lexing speed. This needs a speed test before/after this change. I suggest using Phobos as the source data to use.

benjones committed:
address walters review

benjones commented:
On my 5 year old MBP, it looks like this causes a slight slowdown on average.  I'll try to profile it and see if I can spot any obvious issues.  I assumed my division/modulo operations would be optimized to bitwise ops... does DMD do that?

Results for running make clean then time make (3 cycles, the first of which will rebuild druntime)
master:
27.50s, 17.67s, 17.66s
this PR with poolsize == 1024:
26.69s, 18.28s, 18.04s
This PR with poolsize = 256:
26.17s, 18.20s, 17.81s
This PR with poolsize = 64:
26.53s, 18.07s, 18.10s

PS = 256 seemed like a good tradeoff, although the speed doesn't seem to be too sensitive

benjones committed:
use a plain Array for allocating tokens

benjones commented:
Just using an Array is simpler code that's as fast or faster as the pool approach I was using.  It doesn't seem like anyone was relying on the address of a Token remaining constant

wilzbach commented:
> On my 5 year old MBP, it looks like this causes a slight slowdown on average. I'll try to profile it and see if I can spot any obvious issues. I assumed my division/modulo operations would be optimized to bitwise ops...

I highly recommend only running your benchmark with LDC. DMD compiled with LDC is easily twice as fast (for math code it's even more), so I don't see why we should cater for DMD compiled with DMD for any kind of performance comparison and analysis.
In Phobos, for example, often refactorings result in slower code with DMD as it can't inline as much anymore, but we have a long list of precedents to accept them nevertheless if the performance with LDC doesn't degrade.

benjones commented:
What's the easiest way to compile dmd with ldc?  `HOST_DMD=/my/ldc2 make -f posix.mak` gave me an error about the -vtls flag

MoonlightSentinel commented:
@benjones Use the ldc wrapper `ldmd2`, it converts DMDs command line flags for ldc

benjones commented:
timing make -f posix.mak in phobos after a clean, using dmd built with HOST_DMD=ldmd2, I get pretty much the same results.
First of each rebuilds druntime
master: 25.75, 17.83, 17.36
latest PR: 26.36, 18.10, 17.71

Running both in the profiler, very little time is spent in the lexer in either version

MoonlightSentinel commented:
Did you use `ENABLE_RELEASE=1` and `ENABLE_LTO=1`?

benjones commented:
Nope.  Take 3 :)

Results for building dmd with:  ENABLE_RELEASE=1 ENABLE_LTO=1 HOST_DMD=/usr/local/bin/ldmd2 VERBOSE=1 make -f posix.mak

master: 13.40, 9.64, 9.42
PR: 14.05, 9.56, 9.46

performance impact seems to be in the noise

rainers commented:
Slightly offtopic: @CyberShadow can you please resurrect https://perf.dlang.io/ . Last results are from September.

CyberShadow commented:
OK.

Looks like there is a compilation performance regression:

https://perf.dlang.io/#program-hello-compile-callgrind-insns;;1566740471.47766;1576430864

It will bisect towards it, should find the cause in a few hours.

(That, or something changed on the server OS installation to cause it; if it ends up showing 
the "cause" as the commit after f418f447f5fc (2019-09-19 06:36:06) then this would be the reason.)

PetarKirov commented:
@benjones First off, thanks for working on this!
As you may have realized, doing an end-to-end build is not a good benchmark. Instead you should try to run just the just the lexer and parser on D source code.

benjones commented:
@PetarKirov Any suggestions on how best to do that?  To some extent, the benchmarks show that this is more of a QOL improvement since very little of DMDs runtime is spent doing lexing AFAICT

thewilsonator commented:
The thing about lexing is that it is on the critical path, meaning any slowdown will be realised 100% on all compilations.

benjones commented:
@thewilsonator yeah, that makes sense.  Are there any lexing benchmarks around?  Is that something we want to add to the test suite before merging this PR?  

I haven't profiled DMD built with LDC, but for the one built with DMD, from what I could tell in the profiler, only about 100-200ms of the 17s phobos compile was spent in the lexer,

benjones commented:
Now that the holidays are over I can keep working on this.... What should I do to make sure this doesn't introduce performance regressions?

jacob-carlborg commented:
> Are there any lexing benchmarks around?

Here's a unit test for the diagnostic reporter for the lexer [1]. It shows an example how to only use the lexer.

[1] https://github.com/dlang/dmd/blob/master/test/unit/lexer/diagnostic_reporter.d

jacob-carlborg commented on src/dmd/hdrgen.d:
```diff
@@ -717,23 +718,23 @@ public:
+        TokenRange tokens = s.tokens;
         buf.level++;
-        while (t)
+        while (!tokens.empty)
```
Can this be a `foreach` loop instead?

benjones replied:
I think I tried that, but the calls to tokens.peekFront won't work.  They always return the 2nd token in the original range.

jacob-carlborg replied:
Hmm, that's a shame. The reason for using the range API is to be able to use `foreach` and algorithms from `std.algorithm`. If that doesn't work I'm not sure what's the point.

jacob-carlborg commented on src/dmd/lexer.d:
```diff
@@ -202,6 +203,104 @@ unittest
+{
+    private Array!Token tokens;
+
+    public ref Token getByIndex(size_t index) nothrow pure
```
Perhaps use operator overloading instead making the following syntax possible: `pool[index]`.

jacob-carlborg commented on src/dmd/lexer.d:
```diff
@@ -202,6 +203,104 @@ unittest
+ */
+struct TokenRange
+{
+private:
```
I recommend using `private` in front of each instance variable instead and remove `public:`.

jacob-carlborg commented on src/dmd/lexer.d:
```diff
@@ -202,6 +203,104 @@ unittest
+struct TokenRange
+{
+private:
+    Lexer lex; //where are these tokens coming from?
```
`lex` sounds more like a `verb`. I think `lexer` would be a better name.

benjones replied:
I can't rename both the member var and the method to lexer.  Is there a dmd convention for naming "shadowed" members?

jacob-carlborg replied:
I append `_` to the name. Some will prepend it instead.

benjones replied:
Walter told me "Please don't use _ in identifier names unless it is a postfix for a keyword name."

jacob-carlborg replied:
I know. But now there's a reason. Not sure if he prefers `_` to be prepended.

jacob-carlborg commented on src/dmd/lexer.d:
```diff
@@ -202,6 +203,104 @@ unittest
+{
+private:
+    Lexer lex; //where are these tokens coming from?
+    size_t startIndex, stopIndex;
```
Prefer declarations on their own lines.

jacob-carlborg commented on src/dmd/lexer.d:
```diff
@@ -202,6 +203,104 @@ unittest
+        return startIndex == stopIndex;
+    }
+
+    size_t length() const pure @nogc @safe
```
All these methods are undocumented. Even though they are standard range methods they won't show up in the documentation, so it's not possible to know that they're present without looking at the source code.

benjones replied:
Walter told me they didn't need ddoc comments since they're standard, should I just add /// above them?

jacob-carlborg replied:
Yes, I noticed he mentioned that. I think you should add `///`, not sure if he will object.

jacob-carlborg commented on src/dmd/lexer.d:
```diff
@@ -202,6 +203,104 @@ unittest
+    {
+        return startIndex;
+    }
+    /// Returns: the index of one past the last token, in the lexer's array of tokens
```
I suggest a blank line between the closing curly brace and the documentation.

jacob-carlborg commented on src/dmd/lexer.d:
```diff
@@ -300,62 +403,60 @@ class Lexer
+    }
+
+    ///make a range containing only this.token
+    final TokenRange makeRangeFromHere()
```
Perhaps `currentTokenRange` would be a better name?

benjones replied:
I like the idea of opslice as there's a couple places where I advance it immediately.  My worry is that it might be a little confusing what the indices mean, but I supposed making them relative to the current token makes sense.  Good suggestion!

jacob-carlborg replied:
You don't need any indexes. Just declare it as `final TokenRange opSlice()`.

jacob-carlborg commented on src/dmd/lexer.d:
```diff
@@ -202,6 +203,104 @@ unittest
+        return lex.getByIndex(stopIndex++);
+    }
+    /// Returns: the lexer this range gets its tokens from
+    Lexer getLexer()
```
This should only be called `lexer`, to make it more like a property.

jacob-carlborg commented on src/dmd/lexer.d:
```diff
@@ -202,6 +203,104 @@ unittest
+struct TokenRange
+{
+private:
+    Lexer lex; //where are these tokens coming from?
```
BTW, is there a need to make this private since there's a getter available?

benjones replied:
I don't think you should be able to change which lexer the range is attached to.  It's basically logically const, but the methods I call might affect the lexer's "read ahead buffer" so it can't be actually const.

jacob-carlborg replied:
Fair enough.

jacob-carlborg commented on src/dmd/lexer.d:
```diff
@@ -300,62 +403,60 @@ class Lexer
+
+    }
+
+    final size_t currentPosition()
```
`const pure @nogc @safe`

jacob-carlborg commented on src/dmd/lexer.d:
```diff
@@ -300,62 +403,60 @@ class Lexer
+    }
+
+    ///make a range containing only this.token
+    final TokenRange makeRangeFromHere()
```
`const pure @nogc @safe`

benjones committed:
address @jacob-carlborg review

benjones commented:
In trying to understand why my version was spending so much more time in malloc than master, I realized that I had fundamentally misunderstood how the current implementation works.   I had assumed that the entire token stream stays in memory as a linked list, when really only the currently "peeked" tokens live simultaneously and otherwise the current token just gets repeatedly overwritten...

Parts of this might still be useful, but I think this is probably no good in its current state.  I'll take a look at what's salvageable.
